diff --git a/src/memray/_vendor/textual/__init__.py b/src/memray/_vendor/textual/__init__.py
--- a/src/memray/_vendor/textual/__init__.py
+++ b/src/memray/_vendor/textual/__init__.py
@@ -33,23 +33,10 @@
 LogCallable: TypeAlias = "Callable"
 
 
-if TYPE_CHECKING:
-    from importlib.metadata import version
-
-    from textual.app import App as _App
-
-    __version__ = version("textual")
-    """The version of Textual."""
-
-else:
-
-    def __getattr__(name: str) -> str:
-        """Lazily get the version."""
-        if name == "__version__":
-            from importlib.metadata import version
-
-            return version("textual")
-        raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+__version__ = "8.2.1"
+
+if TYPE_CHECKING:
+    from textual.app import App as _App
 
 
 class LoggerError(Exception):
diff --git a/src/memray/_vendor/textual/widgets/__init__.py b/src/memray/_vendor/textual/widgets/__init__.py
--- a/src/memray/_vendor/textual/widgets/__init__.py
+++ b/src/memray/_vendor/textual/widgets/__init__.py
@@ -106,10 +106,14 @@ def __getattr__(widget_class: str) -> type[Widget]:
         pass
 
     if widget_class not in __all__:
-        raise AttributeError(f"Package 'textual.widgets' has no class '{widget_class}'")
+        raise AttributeError(
+            f"Package 'memray._vendor.textual.widgets' has no class '{widget_class}'"
+        )
 
     widget_module_path = f"._{camel_to_snake(widget_class)}"
-    module = import_module(widget_module_path, package="textual.widgets")
+    module = import_module(
+        widget_module_path, package="memray._vendor.textual.widgets"
+    )
     class_ = getattr(module, widget_class)
 
     _WIDGETS_LAZY_LOADING_CACHE[widget_class] = class_
